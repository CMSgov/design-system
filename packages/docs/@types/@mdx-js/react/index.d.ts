declare module '@mdx-js/react' {
  const MDXProvider: React.ComponentType;
  export { MDXProvider };
}
