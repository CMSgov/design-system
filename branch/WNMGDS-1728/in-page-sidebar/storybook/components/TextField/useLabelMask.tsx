import React, { useState, useRef } from 'react';
import classNames from 'classnames';
import uniqueId from 'lodash/uniqueId';
import { TextInputProps } from './TextInput';

export type MaskFunction = (rawInput: string, valueOnly?: boolean) => string;

export const RE_DATE = /^(\d{1,2})[\D]?(\d{1,2})?[\D]?(\d{1,4})?/;

/**
 * Takes the string value from an input and returns a string
 * with appropriate date format masking applied concatenated
 * with the date hint text, MM/DD/YYYY.
 *
 * `valueOnly` defaults to false. If true, returns formatted
 * string without additional hint text.
 */
export const DATE_MASK: MaskFunction = (rawInput = '', valueOnly = false) => {
  const match = RE_DATE.exec(rawInput);
  let formattedDate = '';
  if (match) {
    const [month, day, year] = match.slice(1);
    formattedDate = [
      // We treat all non-numeric characters as a delimiter. If they're using a
      // delimiter after a month or day, we interpret that as the user supplying
      // a single digit for month or day, which we will automatically pad for them.
      month && month.padStart(2, '0'),
      day && day.padStart(2, '0'),
      year,
    ]
      .filter((s) => s)
      .join('/');
  }
  if (valueOnly) {
    return formattedDate;
  }
  const hint = 'MM/DD/YYYY';
  const hintSub = hint.substring(formattedDate.length);
  return formattedDate + hintSub;
};

export function useLabelMask(maskFn: MaskFunction, originalInputProps: TextInputProps) {
  // TODO: Once we're on React 18, we can use the `useId` hook
  const labelMaskId = useRef(uniqueId('labelmask_')).current;
  const [focused, setFocused] = useState(false);
  const { onFocus, onBlur, onChange } = originalInputProps;
  const value = originalInputProps.value?.toString() ?? '';

  const inputProps = {
    ...originalInputProps,
    defaultValue: undefined,
    onFocus: (e) => {
      if (onFocus) {
        onFocus(e);
      }

      setFocused(true);
    },
    onBlur: (e) => {
      const maskedValue = maskFn(value.toString(), true);
      e.currentTarget.value = maskedValue;
      e.target.value = maskedValue;

      if (onChange) {
        (onChange as any)(e);
      }

      if (onBlur) {
        (onBlur as any)(e);
      }

      setFocused(false);
    },
    type: 'text',
    inputMode: 'numeric' as const,
    'aria-describedby': classNames(originalInputProps['aria-describedby'], labelMaskId),
  };

  return {
    labelMask: (
      <div className="ds-c-label-mask" id={labelMaskId}>
        <span className={classNames(focused && 'ds-u-visibility--screen-reader')}>
          {maskFn('')}
        </span>
        <span className={classNames(!focused && 'ds-u-display--none')} aria-hidden="true">
          {maskFn(value)}
        </span>
      </div>
    ),
    inputProps,
  };
}

export default useLabelMask;
