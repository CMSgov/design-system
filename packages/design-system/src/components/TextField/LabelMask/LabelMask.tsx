import React from 'react';

function maskDate(rawInput = '', valueOnly = false): any {
  const RE_DATE = /^(\d{1,2})[\D]?(\d{1,2})?[\D]?(\d{1,4})?/;
  const match = RE_DATE.exec(rawInput);
  let formattedDate = '';
  if (match) {
    const [month, day, year] = match.slice(1);
    formattedDate = [
      // We treat all non-numeric characters as a delimiter. If they're using a
      // delimiter after a month or day, we interpret that as the user supplying
      // a single digit for month or day, which we will automatically pad for them.
      month && month.padStart(2, '0'),
      day && day.padStart(2, '0'),
      year,
    ]
      .filter((s) => s)
      .join('/');
  }
  if (valueOnly) {
    return formattedDate;
  }
  const hint = 'MM/DD/YYYY';
  const hintSub = hint.substring(formattedDate.length);
  return formattedDate + hintSub;
}

export const BuiltInMask = {
  DATE: maskDate,
};

export interface LabelMaskProps {
  /**
   * Must contain a `TextField` component
   */
  children: React.ReactNode;
  labelMask?: (rawInput: string, valueOnly?: boolean) => string;
}

const LabelMask = ({ children, labelMask }: LabelMaskProps) => {
  if (typeof labelMask === 'string') {
    if (labelMask === BuiltInMask.DATE) {
      labelMask = maskDate;
    }
  }

  /**
   * Get the child text field. Called as a method so that
   * updates to the field cause the mask to re-render
   * @returns {React.ReactElement} Child TextField
   */
  const field = (): React.ReactElement => React.Children.only(children as React.ReactElement);
  const value = field().props.value;

  /**
   * @param {Object} evt
   * @param {React.Element} field - Child TextField
   */
  const handleChange = (evt: React.ChangeEvent<HTMLInputElement>) => {
    if (typeof field().props.onChange === 'function') {
      return field().props.onChange(evt);
    }
  };

  const modifiedTextField = React.cloneElement(field(), {
    defaultValue: undefined,
    onChange: (e) => handleChange(e),
    onBlur: (e) => {
      const maskedValue = labelMask(value, true);
      e.currentTarget.value = maskedValue;
      e.target.value = maskedValue;
      handleChange(e);

      if (typeof field().props.onBlur === 'function') {
        return field().props.onBlur(e);
      }
    },
    type: 'text',
    inputMode: 'numeric',
  });

  return (
    <div className="ds-c-label-mask">
      {modifiedTextField}
      <div className="ds-c-label-mask__mask" aria-hidden="true">
        <span className="ds-c-label-mask__mask--active">{labelMask(value)}</span>
        <span className="ds-c-label-mask__mask--inactive">{labelMask('')}</span>
      </div>
    </div>
  );
};

export default LabelMask;
