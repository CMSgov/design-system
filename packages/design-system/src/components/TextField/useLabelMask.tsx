import React from 'react';
import { useEffect, useState } from 'react';
import classNames from 'classnames';
import { TextInputProps } from './TextInput';
import mergeIds from '../utilities/mergeIds';
import useId from '../utilities/useId';

export type MaskFunction = (rawInput: string, valueOnly?: boolean) => string;

/**
 * Regular expressions for matching commonly used masks, ensure capture groups are
 * set for each set of items that are separated by a delimiter, delimiters are
 * optional and so should be shown as well with the `?` optional flag.
 */
export const RE_DATE = /^(\d{1,2})[\D]?(\d{1,2})?[\D]?(\d{1,4})?/;
export const RE_PHONE = /^\(?(\d{1,3})?\)?[\s.-]?(\d{1,3})?[\s.-]?(\d{1,4})?/;
export const RE_SSN = /([*\d]{1,3})[\s.-]?([*\d]{1,2})?[\s.-]?([\d{1,4}]+)?/;
export const RE_ZIP = /(\d{1,5})/;
export const RE_CURRENCY = /\$?[\d,.-]*/;

/**
 * This function returns a mask function which returns either the formatted match only, or
 * the formatted match + the hint substring which the regular expressions above match against
 * when determining what content has been filled in.
 */
const makeMask = (regex: RegExp, hint: string, formatter: (stringMatch: string[]) => string) => {
  return (rawInput = '', valueOnly = false) => {
    const match = regex.exec(rawInput);
    const formattedMatch = match ? formatter(match) : '';

    if (valueOnly) {
      return formattedMatch;
    }

    const hintSub = hint.substring(formattedMatch.length);
    return formattedMatch + hintSub;
  };
};

/**
 * The date mask automatically pads months and days that are one digit
 */
export const DATE_MASK: MaskFunction = makeMask(RE_DATE, 'MM/DD/YYYY', (match) => {
  const [month, day, year] = match.slice(1);
  const formattedDate = [
    // We treat all non-numeric characters as a delimiter. If they're using a
    // delimiter after a month or day, we interpret that as the user supplying
    // a single digit for month or day, which we will automatically pad for them.
    month && month.padStart(2, '0'),
    day && day.padStart(2, '0'),
    year,
  ]
    .filter((s) => s)
    .join('/');

  return formattedDate;
});

/**
 * Formatting for US phone numbers
 */
export const PHONE_MASK: MaskFunction = makeMask(RE_PHONE, '###-###-####', (match) => {
  return match
    .slice(1)
    .filter((s) => s)
    .join('-');
});

/**
 * Formatting for US Postal codes, this could be expanded to allow for 9 digit numbers
 */
export const ZIP_MASK: MaskFunction = makeMask(RE_ZIP, '#####', (match) => {
  return match[1];
});

/**
 * Formatting for social security numbers.
 */
export const SSN_MASK: MaskFunction = makeMask(RE_SSN, '###-##-####', (match) => {
  /**
   * for future notice, there are rules regarding social security numbers, these could be validated
   * if we were to add error handling to useLabelMask as an additional set of functionality
   * The first part should have 3 digits and should not be 000, 666, or between 900 and 999.
   * The second part should have 2 digits and it should be from 01 to 99.
   * The third part should have 4 digits and it should be from 0001 to 9999.
   */
  return match
    .slice(1)
    .filter((s) => s)
    .join('-');
});

/**
 * Does the same thing as SSN_MASK except that it obfuscates the first five digits
 */
export const SSN_MASK_OBFUSCATED: MaskFunction = (rawInput: string, valueOnly?: boolean) => {
  // Use the normal SSN_MASK function just to clean the raw input and format it
  const formatted = SSN_MASK(rawInput, true);
  // We only hide the first five digits of the SSNs
  const obfuscation = '***-**';

  let obfuscated: string;
  if (formatted.length < obfuscation.length) {
    obfuscated = obfuscation.substring(0, formatted.length);
  } else {
    obfuscated = obfuscation + formatted.substring(obfuscation.length);
  }

  return SSN_MASK(obfuscated, valueOnly);
};

/**
 * Currency mask is a little different, we need to modify the incoming content to strip
 * out any commas or dollar signs before evaluating it via the Intl.NumberFormat function.
 */
export const CURRENCY_MASK = makeMask(RE_CURRENCY, '$', (match) => {
  const signed = match[0].includes('-');
  const stripped = match[0].replace(/[^0-9.]/g, '');
  const clipped = stripped.includes('.') ? stripped.slice(0, stripped.indexOf('.') + 3) : stripped;
  const USDollar = new Intl.NumberFormat('en-US', { style: 'currency', currency: 'USD' });
  const formatted = USDollar.format(Number(clipped)).replace(/\.00/, '');

  if (Number(clipped) > 0) {
    return signed ? '-' + formatted : formatted;
  } else {
    return '';
  }
});

export function useLabelMask(maskFn: MaskFunction, originalInputProps: TextInputProps) {
  // TODO: Once we're on React 18, we can use the `useId` hook
  const generatedId = useId('label-mask--');
  const labelMaskId = originalInputProps.id ? `${originalInputProps.id}__label-mask` : generatedId;
  const [focused, setFocused] = useState(false);
  const { onFocus, onBlur, onChange } = originalInputProps;
  const value =
    originalInputProps.value?.toString() ?? originalInputProps.defaultValue?.toString() ?? '';
  const [currentValue, setCurrentValue] = useState(value);

  useEffect(() => {
    setCurrentValue(value);
  }, [value, setCurrentValue]);

  const inputProps = {
    ...originalInputProps,
    defaultValue: undefined,
    value: currentValue,
    onChange: (e: React.ChangeEvent<HTMLInputElement>) => {
      setCurrentValue(e.currentTarget.value);

      if (onChange) {
        onChange(e);
      }
    },
    onFocus: (e: React.FocusEvent<HTMLInputElement>) => {
      if (onFocus) {
        onFocus(e);
      }

      setFocused(true);
    },
    onBlur: (e: React.FocusEvent<HTMLInputElement>) => {
      const maskedValue = maskFn(currentValue, true);
      e.currentTarget.value = maskedValue;
      e.target.value = maskedValue;
      setCurrentValue(maskedValue);

      if (onChange) {
        (onChange as any)(e);
      }

      if (onBlur) {
        (onBlur as any)(e);
      }

      setFocused(false);
    },
    type: 'text',
    inputMode: 'numeric' as const,
    'aria-describedby': mergeIds(originalInputProps['aria-describedby'], labelMaskId),
  };

  let currentMask = maskFn(currentValue);

  // Date mask needs to return the default empty mask when not focused
  if (maskFn === DATE_MASK && !focused) {
    currentMask = maskFn('');
  }

  // SSN mask needs to obfuscate the SSN when not focused
  if (maskFn === SSN_MASK && !focused && currentValue !== '') {
    currentMask = SSN_MASK_OBFUSCATED(currentValue);
    inputProps.value = SSN_MASK_OBFUSCATED(currentValue, true);
  }

  return {
    labelMask: (
      <div className="ds-c-label-mask" id={labelMaskId}>
        <span className={classNames(focused && 'ds-u-visibility--screen-reader')}>
          {currentMask}
        </span>
        <span className={classNames(!focused && 'ds-u-display--none')} aria-hidden="true">
          {currentMask}
        </span>
      </div>
    ),
    inputProps,
  };
}

export default useLabelMask;
