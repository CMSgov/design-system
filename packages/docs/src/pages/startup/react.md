---
title: Using React components
weight: 3
---

The examples below assume you've installed the design system using NPM and have already setup your build system.

<h4 class="ds-h4 ds-u-font-size--base">Default imports</h4>

Individual components can be imported from their individual export file.

```jsx
import Button from '@cmsgov/design-system-core/dist/components/Button/Button';
```

<h4 class="ds-h4 ds-u-font-size--base">Named imports</h4>

Components can also be imported using the shorter syntax below.

<div class="ds-c-alert ds-c-alert--warn ds-u-margin-bottom--2 ds-u-font-size--small">
  <div class="ds-c-alert__body">
    <h3 class="ds-c-alert__heading ds-u-font-size--base">Performance note</h3>
    <p class="ds-c-alert__text">
      This approach may result in a much larger file than you intend. Depending on what module bundler you use, all of the design system's React components may be included in the bundled file even if you didn't specifically import them. This can be avoided by enabling features like [tree shaking in Webpack](https://webpack.js.org/guides/tree-shaking/).
    </p>
  </div>
</div>

```jsx
import { Button } from '@cmsgov/design-system-core';
```

[View an example](https://github.com/CMSgov/design-system/blob/master/examples/react-app/src/scripts/index.js)

<h2 id="examples" class="ds-h2 ds-u-color--primary-darker">Examples</h2>

Additional examples of the design system in use can be viewed on GitHub. These projects demonstrate the various ways you can incorporate the design system into your development process and various use cases.

<a href="https://github.com/CMSgov/design-system/tree/master/examples/" class="ds-c-button">Browse example projects</a>

<h2 id="need-help" class="ds-h2 ds-u-color--primary-darker">Need help or ran into an issue?</h2>

If you're having trouble installing or setting up the design system, or if you think you've found a bug, feel free to [open an issue on GitHub](https://github.com/CMSgov/design-system/issues).
