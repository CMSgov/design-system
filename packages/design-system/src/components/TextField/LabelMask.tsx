import { Children, cloneElement } from 'react';
import type * as React from 'react';
import { useLabelMask, MaskFunction } from './useLabelMask';

export interface LabelMaskProps {
  /**
   * Must contain a `TextField` component
   */
  children: React.ReactNode;
  /**
   * Providing a function here will turn the text field into a label-masked field, where
   * the user input is formatted in a label as the user types and then the input field
   * itself is automatically formatted when the user shifts focus away from the input.
   * A custom function can be given, or one of the following built-in functions can be
   * imported from the design system and passed to this component: PHONE_MASK, ZIP_MASK,
   * SSN_MASK, and CURRENCY_MASK. See
   * [Label-masked field](https://design.cms.gov/components/text-field/label-masked-field/)
   * documentation page for more information.
   */
  labelMask?: MaskFunction;
}

const LabelMask = (props: LabelMaskProps) => {
  const field = Children.only(props.children as React.ReactElement);
  const { labelMask, inputProps } = useLabelMask(props.labelMask, field.props);
  const input = cloneElement(field, inputProps);

  return (
    <>
      {labelMask}
      {input}
    </>
  );
};

export default LabelMask;
