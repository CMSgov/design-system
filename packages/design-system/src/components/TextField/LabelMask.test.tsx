import { useState } from 'react';
import LabelMask from './LabelMask';
import { DATE_MASK, MaskFunction } from './useLabelMask';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

const TestLabelMask = (props: { mask: MaskFunction }) => {
  const { mask } = props;
  const [value, setValue] = useState('');
  return (
    <LabelMask labelMask={mask}>
      <input type="text" value={value} onChange={(e) => setValue(e.currentTarget.value)} />
    </LabelMask>
  );
};

describe('DATE_MASK Label mask', function () {
  it('renders default date mask, MM/DD/YYYY, when no input value set', () => {
    const { container, asFragment } = render(<TestLabelMask mask={DATE_MASK} />);
    const mask = container.querySelector('.ds-c-label-mask');
    const input = container.querySelector('input');

    expect(input.value).toBe('');
    expect(mask.textContent).toContain('MM/DD/YYYY');

    expect(asFragment()).toMatchSnapshot();
  });
  describe('updates label mask to reflect', () => {
    it('complete input value set', () => {
      const { container } = render(<TestLabelMask mask={DATE_MASK} />);
      const data = '12345678';
      const maskText = '12/34/5678';

      const mask = container.querySelector('.ds-c-label-mask');
      const input = container.querySelector('input');

      userEvent.type(input, data);

      expect(input).toHaveValue(data);
      expect(mask.textContent).toContain(maskText);
    });
    it('partial input value set', () => {
      const { container } = render(<TestLabelMask mask={DATE_MASK} />);
      const data = '1234';

      const mask = container.querySelector('.ds-c-label-mask');
      const input = container.querySelector('input');

      userEvent.type(input, data);

      expect(input).toHaveValue(data);
      expect(mask.textContent).toContain('12/34/YYYY');
    });
    it('padded values when incomplete input value set', () => {
      const { container } = render(<TestLabelMask mask={DATE_MASK} />);
      const data = '1';
      const maskText = '01/DD/YYYY';

      const mask = container.querySelector('.ds-c-label-mask');
      const input = container.querySelector('input');

      userEvent.type(input, data);

      expect(input).toHaveValue(data);
      expect(mask.textContent).toContain(maskText);
    });
  });
  it('formats input value onBlur', () => {
    const { container } = render(<TestLabelMask mask={DATE_MASK} />);
    const data = '12345678';
    const formattedData = '12/34/5678';

    const input = container.querySelector('input');

    userEvent.type(input, data);
    userEvent.tab();

    expect(input).toHaveValue(formattedData);
  });
});
