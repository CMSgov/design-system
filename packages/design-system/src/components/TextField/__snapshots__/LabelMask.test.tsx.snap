// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DATE_MASK Label mask renders default date mask, MM/DD/YYYY, when no input value set 1`] = `
<DocumentFragment>
  <div
    class="ds-c-label-mask"
    id="labelmask_1"
  >
    <span
      class=""
    >
      MM/DD/YYYY
    </span>
    <span
      aria-hidden="true"
      class="ds-u-display--none"
    >
      MM/DD/YYYY
    </span>
  </div>
  <input
    aria-describedby="labelmask_1"
    inputmode="numeric"
    type="text"
    value=""
  />
</DocumentFragment>
`;
